python/lsst/butler/butler.py

```diff
@@ -0,0 +1,197 @@
+
+        Returns
+        -------
+        inMemoryDataset : `InMemoryDataset`
+            The requested `Dataset`.
+        """
+        assert isinstance(handle, DatasetHandle)
+        parent = self.datastore.get(handle.uri, handle.type.storageClass, parameters) if handle.uri else None
+        children = {name: self.datastore.get(childHandle, parameters)
+                    for name, childHandle in handle.components.items()}
+        return handle.type.storageClass.assemble(parent, children)
+
+    def get(self, label, parameters=None):
+        """Load a `Dataset` or a slice thereof from the Butler's `Collection`.
+
+        Parameters
+        ----------
+        label : `DatasetLabel`
+            Identifies the `Dataset` to retrieve.
+        parameters : `dict`
+            A dictionary of `StorageClass`-specific parameters that can be used to obtain a slice of the `Dataset`.
+
+        Returns
+        -------
+        dataset : `InMemoryDataset`
+            The requested `Dataset`.
+        """
+        assert isinstance(label, DatasetLabel)
+        handle = self.registry.find(self.config.collection, label)
+        if handle:
+            return self.getDirect(handle, parameters)
+        else:
+            return None  # No Dataset found
+
+    def put(self, label, inMemoryDataset, producer=None):
+        """Write a `Dataset`.
+
+        Parameters
+        ----------
+        label : `DatasetLabel`
+            Identifies the `Dataset` being stored.
+        inMemoryDataset : `InMemoryDataset`
+            The `Dataset` to store.
+        producer : `Quantum`
+            Identifies the producer of this `Dataset`.  May be ``None`` for some `Registries`.
+            ``producer.run`` must match ``self.config.run``.
+
+        Returns
+        -------
+        datasetHandle : `DatasetHandle`
+            A handle that identifies the registered (and stored) dataset.
+        """
+        ref = self.registry.expand(label)
+        run = self.config.run
+        assert(producer is None or run == producer.run)
+        # template = self.config.templates.get(ref.type.name, None)
+        # path = ref.makePath(run, template)
+        path = "temp.fits"  # TODO: Implement this!
+        uri, components = self.datastore.put(inMemoryDataset, ref.type.storageClass, path, ref.type.name)
+        return self.registry.addDataset(ref, uri, components, producer=producer, run=run)
+
+    def markInputUsed(self, quantum, ref):
+        """Mark a `Dataset` as having been "actually" (not just predicted-to-be) used by a `Quantum`.
+
+        Parameters
+        ----------
+        quantum : `Quantum`
+            The dependent `Quantum`.
+        ref : `DatasetRef`
+            The `Dataset` that is a true dependency of ``quantum``.
+        """
+        handle = self.registry.find(self.config.collection, ref)
+        self.registry.markInputUsed(handle, quantum)
+
+    def unlink(self, *labels):
```

Comment:
What does `unlink` mean for a read-only Butler?

Reply:
Nothing. Should not be implemented in this case, or be a noop (yet to decide).